### ISSUE TITLE
feature request: support for GCS

### ISSUE THREAD

**Issue by rbtcollins** [j3t/jenkins-pipeline-cache-plugin#26]
Hi, after https://github.com/jenkinsci/jobcacher-plugin/issues/137 we decided to try jenkins-pipeline-cache-plugin, but similar result (though more details here).

We're using GCS with the compatibility layer as described in the issue on the other plugin. Cache upload and download seemed to work ok, but the last-access-time update is failing. Is that a hard dependency, logically speaking, or could it be made optional?

<details>
<summary>
Backtrace com.amazonaws.services.s3.model.AmazonS3Exception: Multiple HTTP header values where one was expected.
</summary>

com.amazonaws.services.s3.model.AmazonS3Exception: Multiple HTTP header values where one was expected. (Service: Amazon S3; Status Code: 400; Error Code: ExcessHeaderValues; Request ID: null; S3 Extended Request ID: null; Proxy: null), S3 Extended Request ID: null
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1879)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1418)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1387)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1157)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:814)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:781)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:755)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:715)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:697)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:561)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:541)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5456)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5403)
	at com.amazonaws.services.s3.AmazonS3Client.copyObject(AmazonS3Client.java:2078)
	at io.jenkins.plugins.pipeline.cache.s3.CacheItemRepository.updateLastAccess(CacheItemRepository.java:102)
	at io.jenkins.plugins.pipeline.cache.agent.RestoreCallable.invoke(RestoreCallable.java:54)
	at io.jenkins.plugins.pipeline.cache.agent.RestoreCallable.invoke(RestoreCallable.java:18)
	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3492)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:377)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:125)
	at java.base/java.lang.Thread.run(Thread.java:833)

</details>

**Comment by j3t** [comment]
Hi, according to the stacktrace, it seems that GCS does not support metadata updates. The last access time is stored as metadata value but since last modified and last access are equal (metadata updates also changing last modified) we could get rid of it and use last modified instead. I will take a look, it should be possible.

**Comment by j3t** [comment]
@rbtcollins I'm wondering, is GCS not supporting metadata values at all or just not metadata updates? Since the upload of a new cache item seems to work fine, I would also expect that the initial metadata values are present (`CREATION` and `LAST_ACCESS`). Can you check that please?

**Comment by rbtcollins** [comment]
Custom metadata seems well supported.
![image](https://user-images.githubusercontent.com/499678/211479453-745607ce-e1c2-457c-86f2-960f5d104787.png)


The error is complaining about a repeated header, but as I can't see the full request I can't tell what has been repeated to guess at the cause.

**Comment by rbtcollins** [comment]
I'm wondering if https://github.com/j3t/jenkins-pipeline-cache-plugin/blob/main/src/main/java/io/jenkins/plugins/pipeline/cache/s3/CacheItemRepository.java#L99 actually ends up with two LAST_ACCESSED custom metadata headers, and then AWS is discarding one. Perhaps delete the metadata entry then add it, and it would fix the issue?

**Comment by rbtcollins** [comment]
Captured a log of the failure. 


http-outgoing-4 >> "PUT /<bucket>/infrastructure-d41d8cd98f00b204e9800998ecf8427e HTTP/1.1[\r][\n]"
http-outgoing-4 >> "Host: storage.googleapis.com[\r][\n]"
http-outgoing-4 >> "Accept-Ranges: bytes[\r][\n]"
http-outgoing-4 >> "amz-sdk-invocation-id: fd0f1165-ecaf-b259-0012-ea122e8f4f87[\r][\n]"
http-outgoing-4 >> "amz-sdk-request: ttl=20230110T084128Z;attempt=1;max=4[\r][\n]"
http-outgoing-4 >> "amz-sdk-retry: 0/0/500[\r][\n]"
http-outgoing-4 >> "Authorization: AWS4-HMAC-SHA256 Credential=XXXX/s3/aws4_request, SignedHeaders=accept-ranges;amz-sdk-invocation-id;amz-sdk-request;amz-sdk-retry;cache-control;content-length;content-type;etag;expires;host;last-modified;user-agent;x-amz-content-sha256;x-amz-copy-source;x-amz-date;x-amz-meta-creation;x-amz-meta-last_access;x-amz-metadata-directive;x-goog-expiration;x-goog-generation;x-goog-hash;x-goog-metageneration;x-goog-storage-class;x-goog-stored-content-encoding;x-goog-stored-content-length;x-guploader-uploadid, Signature=593cc06fb5c9bdaf7a04e976d71f9575cb88a8c5701c34d87d93b068643a0fb2[\r][\n]"
http-outgoing-4 >> "Cache-Control: private, max-age=0[\r][\n]"
http-outgoing-4 >> "Content-Type: application/octet-stream[\r][\n]"
http-outgoing-4 >> "ETag: 5edeca4638f31588e54433be43035416[\r][\n]"
http-outgoing-4 >> "Expires: Tue, 10 Jan 2023 08:40:38 GMT[\r][\n]"
http-outgoing-4 >> "Last-Modified: Mon Jan 09 21:24:09 UTC 2023[\r][\n]"
http-outgoing-4 >> "User-Agent: aws-sdk-java/1.12.287 Linux/5.10.127+ OpenJDK_64-Bit_Server_VM/17.0.4+8-Ubuntu-120.04 java/17.0.4 groovy/2.4.21 vendor/Private_Build cfg/retry-mode/legacy[\r][\n]"
http-outgoing-4 >> "x-amz-content-sha256: UNSIGNED-PAYLOAD[\r][\n]"
http-outgoing-4 >> "x-amz-copy-source: /<bucket>/infrastructure-d41d8cd98f00b204e9800998ecf8427e[\r][\n]"
http-outgoing-4 >> "X-Amz-Date: 20230110T084038Z[\r][\n]"
http-outgoing-4 >> "x-amz-meta-creation: 1673299449275[\r][\n]"
http-outgoing-4 >> "x-amz-meta-last_access: 1673340038385[\r][\n]"
http-outgoing-4 >> "x-amz-metadata-directive: REPLACE[\r][\n]"
http-outgoing-4 >> "x-goog-expiration: Sat, 08 Jul 2023 21:24:09 GMT[\r][\n]"
http-outgoing-4 >> "x-goog-generation: 1673299449357909[\r][\n]"
http-outgoing-4 >> "x-goog-hash: md5=Xt7KRjjzFYjlRDO+QwNUFg==[\r][\n]"
http-outgoing-4 >> "x-goog-metageneration: 1[\r][\n]"
http-outgoing-4 >> "x-goog-storage-class: REGIONAL[\r][\n]"
http-outgoing-4 >> "x-goog-stored-content-encoding: identity[\r][\n]"
http-outgoing-4 >> "x-goog-stored-content-length: 2048[\r][\n]"
http-outgoing-4 >> "X-GUploader-UploadID: ADPycds918H3d1Sz4sqeIiBjh4F8orzpHE7zob6jn8aKq-yVhr39vcsKU6q-KERV3yen3IdfPWPII3ipFgLFbsrDmMi7AV3CPcws[\r][\n]"
http-outgoing-4 >> "Content-Length: 0[\r][\n]"
http-outgoing-4 >> "Connection: Keep-Alive[\r][\n]"
http-outgoing-4 >> "[\r][\n]"


http-outgoing-4 << "HTTP/1.1 400 Bad Request[\r][\n]"
http-outgoing-4 << "X-GUploader-UploadID: ADPycdsfmBsMh4mL3dmNu3-mrZLx14CqeNisExdt3iFQ5bRgsvsK8ZfG0syYdbPkmCzIFO9IK2u3W4_xCIEc6ODkzEdPWw[\r][\n]"
http-outgoing-4 << "Content-Type: application/xml; charset=UTF-8[\r][\n]"
http-outgoing-4 << "Content-Length: 226[\r][\n]"
http-outgoing-4 << "Date: Tue, 10 Jan 2023 08:40:38 GMT[\r][\n]"
http-outgoing-4 << "Server: UploadServer[\r][\n]"
http-outgoing-4 << "[\r][\n]"
http-outgoing-4 << "<?xml version='1.0' encoding='UTF-8'?><Error><Code>ExcessHeaderValues</Code><Message>Multiple HTTP header values where one was expected.</Message><Details>Requests cannot specify both x-amz and x-goog headers</Details></Error>"

**Comment by rbtcollins** [comment]
So whats happening is that the copy of the metadata you use to set new object metadata has googles compatish headers, and then these conflict with the semantics of the AWS SDK added headers. I think. Grabbing the read request before to confirm.

**Comment by rbtcollins** [comment]
And the HEAD request to get the metadata that is then copied. 

<details>
<summary> HEAD request</summary>

```
http-outgoing-4 >> "HEAD /<bucket>/infrastructure-d41d8cd98f00b204e9800998ecf8427e HTTP/1.1[\r][\n]"
http-outgoing-4 >> "Host: storage.googleapis.com[\r][\n]"
http-outgoing-4 >> "amz-sdk-invocation-id: efdf8a54-d332-06f1-d279-89e3124beae5[\r][\n]"
http-outgoing-4 >> "amz-sdk-request: ttl=20230110T084128Z;attempt=1;max=4[\r][\n]"
http-outgoing-4 >> "amz-sdk-retry: 0/0/500[\r][\n]"
http-outgoing-4 >> "Authorization: AWS4-HMAC-SHA256 Credential=XXX/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;amz-sdk-retry;content-type;host;user-agent;x-amz-content-sha256;x-amz-date, Signature=ff16c58233fdd18226417a9158909b0ad1bdea99fa88427451fae8ce95e40b10[\r][\n]"
http-outgoing-4 >> "Content-Type: application/octet-stream[\r][\n]"
http-outgoing-4 >> "User-Agent: aws-sdk-java/1.12.287 Linux/5.10.127+ OpenJDK_64-Bit_Server_VM/17.0.4+8-Ubuntu-120.04 java/17.0.4 groovy/2.4.21 vendor/Private_Build cfg/retry-mode/legacy[\r][\n]"
http-outgoing-4 >> "x-amz-content-sha256: UNSIGNED-PAYLOAD[\r][\n]"
http-outgoing-4 >> "X-Amz-Date: 20230110T084038Z[\r][\n]"
http-outgoing-4 >> "Connection: Keep-Alive[\r][\n]"
http-outgoing-4 >> "[\r][\n]"


http-outgoing-4 << "HTTP/1.1 200 OK[\r][\n]"
http-outgoing-4 << "X-GUploader-UploadID: ADPycds918H3d1Sz4sqeIiBjh4F8orzpHE7zob6jn8aKq-yVhr39vcsKU6q-KERV3yen3IdfPWPII3ipFgLFbsrDmMi7AV3CPcws[\r][\n]"
http-outgoing-4 << "Expires: Tue, 10 Jan 2023 08:40:38 GMT[\r][\n]"
http-outgoing-4 << "Date: Tue, 10 Jan 2023 08:40:38 GMT[\r][\n]"
http-outgoing-4 << "Cache-Control: private, max-age=0[\r][\n]"
http-outgoing-4 << "Last-Modified: Mon, 09 Jan 2023 21:24:09 GMT[\r][\n]"
http-outgoing-4 << "ETag: "5edeca4638f31588e54433be43035416"[\r][\n]"
http-outgoing-4 << "x-goog-generation: 1673299449357909[\r][\n]"
http-outgoing-4 << "x-goog-metageneration: 1[\r][\n]"
http-outgoing-4 << "x-goog-stored-content-encoding: identity[\r][\n]"
http-outgoing-4 << "x-goog-stored-content-length: 2048[\r][\n]"
http-outgoing-4 << "x-amz-meta-creation: 1673299449275[\r][\n]"
http-outgoing-4 << "x-amz-meta-last_access: 1673299449275[\r][\n]"
http-outgoing-4 << "Content-Type: application/octet-stream[\r][\n]"
http-outgoing-4 << "x-goog-hash: crc32c=vRjx8Q==[\r][\n]"
http-outgoing-4 << "x-goog-hash: md5=Xt7KRjjzFYjlRDO+QwNUFg==[\r][\n]"
http-outgoing-4 << "x-goog-expiration: Sat, 08 Jul 2023 21:24:09 GMT[\r][\n]"
http-outgoing-4 << "x-goog-storage-class: REGIONAL[\r][\n]"
http-outgoing-4 << "Accept-Ranges: bytes[\r][\n]"
http-outgoing-4 << "Content-Length: 2048[\r][\n]"
http-outgoing-4 << "Server: UploadServer[\r][\n]"
http-outgoing-4 << "[\r][\n]"
```
</details>

**Comment by rbtcollins** [comment]
So - we know the AWS SDK will set x-amz headers, and the GCS responses are coming back with *additional* x-goog headers. I'm not sure that any of them are needed, since a default copy doesn't set object metadata at all and just copies it over, and the copy request represents metadata as just additional headers.

So I think:
 - don't copy the metadata across
 - make a new metadata object and set the two fields desired explictly.

**Comment by j3t** [comment]
> make a new metadata object and set the two fields desired explictly.

I'm not sure if this solves our problem but we can try it. Could you please check the attached version?
[build artifacts.zip](https://github.com/j3t/jenkins-pipeline-cache-plugin/files/10381838/build.artifacts.zip)

**Comment by j3t** [comment]
should be fixed with version 0.1.6
